### PR TITLE
Resolve #2435: Lucene optimized postings

### LINKED ISSUE #2435: Move Lucene's postings to K/V pairs

### PR DESCRIPTION
**Pull request by ohadzeliger**

### CODE REVIEW

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+import static com.apple.foundationdb.record.metadata.Key.Expressions.field;
+import static com.apple.foundationdb.record.metadata.Key.Expressions.function;
+import static com.apple.foundationdb.record.metadata.Key.Expressions.value;
+import static com.apple.foundationdb.record.provider.foundationdb.indexes.TextIndexTestUtils.COMPLEX_DOC;
+import static com.apple.foundationdb.record.provider.foundationdb.indexes.TextIndexTestUtils.SIMPLE_DOC;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Tests for Lucene StoredFields implementation.
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java:61]:
```suggestion
 * Tests for Lucene Postings implementation.
```

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+        try (FDBRecordContext context = openContext(contextProps)) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            final RecordQuery query = buildQuery("Document", Collections.emptyList(), SIMPLE_DOC);
+            queryAndAssertFields(query, "text", Map.of(1624L, "Document 2"));
+            LuceneIndexTestUtils.mergeSegments(recordStore, index);
+            try (FDBDirectory directory = new FDBDirectory(recordStore.indexSubspace(index), context, index.getOptions())) {
+                // After a merge, all tombstones are removed and one document remains
+                assertFieldCountPerSegment(directory, List.of("_0", "_1", "_2"), List.of(0, 0, 1));
+            }
+            assertTrue(timer.getCounter(LuceneEvents.Counts.LUCENE_DELETE_POSTINGS).getCount() > 0);
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java:196]: This doesn't happen until after the merge, perhaps worth asserting that is the case.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+            context.commit();
+        }
+        try (FDBRecordContext context = openContext(contextProps)) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, index);
+            final RecordQuery query = buildQuery("Document", Collections.emptyList(), SIMPLE_DOC);
+            queryAndAssertFields(query, "text", Map.of());
+            LuceneIndexTestUtils.mergeSegments(recordStore, index);
+            try (FDBDirectory directory = new FDBDirectory(recordStore.indexSubspace(index), context, index.getOptions())) {
+                // When deleting all docs from the index, the first segment (_0) was merged away and the last segment (_1) gets removed
+                assertFieldCountPerSegment(directory, List.of("_0", "_1"), List.of(0, 0));
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java:299]: Why are there two segments left?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+            RecordQuery query = buildQuery("record", Collections.emptyList(), COMPLEX_DOC);
+            // This doc has the PK as (Group, docID)
+            queryAndAssertFieldsTuple(query, "text", Map.of(
+                    Tuple.from(6, 1624L), "Hello record",
+                    Tuple.from(7, 1625L), "Hello record layer"));
+            // query on the other fields
+            query = buildQuery("text2:layer", Collections.emptyList(), COMPLEX_DOC);
+            queryAndAssertFieldsTuple(query, "text2", Map.of(
+                    Tuple.from(7, 1625L), "Hello record layer 2"));
+            query = buildQuery("group:5", Collections.emptyList(), COMPLEX_DOC);
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java:328]: This query and the one below seem irrelevant to the postings format, and could be removed.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+            try (FDBDirectory directory = new FDBDirectory(recordStore.indexSubspace(index), context, index.getOptions())) {
+                // When deleting all docs from the index, the first segment (_0) was merged away and the last segment (_1) gets removed
+                assertFieldCountPerSegment(directory, List.of("_0", "_1"), List.of(0, 0));
+            }
+            assertTrue(timer.getCounter(LuceneEvents.Counts.LUCENE_DELETE_POSTINGS).getCount() > 0);
+        }
+    }
+
+    @Test
+    void testComplexDocManyFields() throws Exception {
```
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePostingsTest.java:306]: I can think of a lot of additional tests that could be beneficial, but given the state of the work, and that this is a temporary pausing point, perhaps it's just worth noting and tracking to come back later.
Mostly, it feels like there's a lot of code that just isn't covered.
1. Tests where positions matter
2. Tests where frequency matters
3. More testing where there are different fields with overlapping terms
4. Tests with terms that may have differences in ordering could mess things up
5. unit test that the ordering is the same between FDB and Lucene terms
6. Tests where impacts matter
7. It might be nice to have some lower-level tests to make sure our api aligns with a built-in implementation, but that may be really hard to get right, because Postings has such a broad api.

**Review by ScottDugas — CHANGES_REQUESTED** [review]

File: docs/ReleaseNotes.md [hunk 6]
```diff
@@ -15,7 +15,7 @@ The Guava dependency version has been updated to 31.1. Projects may need to chec
 // begin next release
 ### NEXT_RELEASE
 
-* **Bug fix** Fix 1 [(Issue #NNN)](https://github.com/FoundationDB/fdb-record-layer/issues/NNN)
+* **Bug fix** Fixe Unbalanced log info exception [(Issue #2392)](https://github.com/FoundationDB/fdb-record-layer/issues/2392)
```
[the remark this answers is not in the document]
Comment (ohadzeliger) [reply docs/ReleaseNotes.md:18]: DOne.

**Comment by ohadzeliger** [comment]
Closed this or as it was open for too long and priorities have changed. Will revisit if necessary.
